### ISSUE TITLE
More Chat Loss Masking Strategies

### ISSUE THREAD

**Issue by EugenHotaj** [pytorch/torchtune#2214]
Are there plans to add more loss masking strategies for chat data? 

E.g. a very common loss masking strategy for multi-turn conversations is to mask everything but the last assistant response. However, `train_on_input=False` right now will compute the loss on all assistant turns, not just the last one. Is it possible to add this feature to torchtune?

**Comment by RdoubleA** [comment]
I just saw a similar request in #2207, so this might be worth enabling

**Comment by EugenHotaj** [comment]
Nice to "see" you again Rafi! Thanks for the quick response.

> But if this is a common approach and other folks would like this for the built in dataset builders then we could consider something like changing train_on_input to a string masking strategy parameter, or something similar.

> I just saw a similar request in https://github.com/pytorch/torchtune/issues/2207, so this might be worth enabling

Masking the last turn only is a very (most?) common masking strategy so could be a nice feature to provide users out of the box.

> If you are using a custom dataset with a custom message transform, you can manually mask the messages you need to in the transform by setting the masked field in the Message dataclass.

Any pointers / examples for how to do this?

**Comment by RdoubleA** [comment]
Glad to see you on the torchtune repo Eugen :)

Yes, see this [page](https://pytorch.org/torchtune/main/basics/message_transforms.html#custom-message-transforms) for an example. 

If your conversation is stored in a column, you can just query that column in the custom message transform and manually create `Message` objects for the whole conversation, leaving the last one unmasked. Then you'll need to make a custom dataset builder that you can specify in your config.

Let me know if there's any confusion on this.

**Comment by tginart** [comment]
Hi @EugenHotaj, is there any example anywhere of how to use train_on_input=False in config?

**Comment by EugenHotaj** [comment]
@tginart I actually ended up hacking my own transformer like @RdoubleA suggested above, it was pretty straightforward:

<img width="537" alt="Screenshot 2025-01-10 at 11 27 18 AM" src="https://github.com/user-attachments/assets/1ab0d8d6-7e5a-4471-a406-f3baa0e6040e" />

**Comment by tginart** [comment]
Ah! Thank you!

 If we just want to use regular masking for user input, how do we turn on ``train_on_input=False`` ? Or is it on by default? @RdoubleA ?

**Comment by RdoubleA** [comment]
By regular masking, do you mean the user input is included in the loss calculation? In this case, you would set train_on_input=True.
```
dataset:
  _component_: torchtune.datasets.alpaca_dataset
  train_on_input: True
```
or add `dataset.train_on_input=True` to your launch command: `tune run full_finetune_single_device --config my_config.yaml dataset.train_on_input=True`

**Comment by ebsmothers** [comment]
@EugenHotaj, @tginart are you guys unblocked on this issue? If so, maybe we can open a separate issue to enumerate and track different masking strategies we want to support. (We can also label as community-help-wanted since I think these would be nice self-contained issues for people to pick up once someone lays out the initial design for custom masking strategies.) Thoughts @RdoubleA?

**Comment by EugenHotaj** [comment]
Closing this one since we're unblocked.
